Add python:3.10 to python versions ci files and some other fixes for fresh installs

meramsey opened the pull request:
Add python:3.10
https://hub.docker.com/layers/python/library/python/3.10/images/sha256-0e6f1cd34a87218544d9e79858dfe68dd0f2d6919c2e2b287983276961aa4942?context=explore

Also some other minor fixes for missing safety dev dependency and mypy failures

```
poetry run duty check
✓ Checking code quality
✗ Type-checking (1)
  > mypy --config-file config/mypy.ini src tests duties.py docs/macros.py
  docs/macros.py:34: error: Argument 1 to "search_packages_info" has incompatible type "Set[Any]"; expected "List[str]"
  docs/macros.py:35: error: Incompatible types in assignment (expression has type "Dict[Any, Union[str, List[str], None]]", variable has type "_PackageInfo")
  docs/macros.py:36: error: Invalid tuple index type (actual type "str", expected type "Union[int, slice]")
  Found 3 errors in 1 file (checked 8 source files)
```

```
@ poetry run duty check
✓ Checking code quality
✓ Type-checking
✓ Building documentation
✗ Checking dependencies (127)
  > poetry export -f requirements.txt --without-hashes | safety check --stdin --full-report
  /usr/bin/sh: 1: safety: not found
  Traceback (most recent call last):
```

meramsey committed:
Update .gitlab-ci.yml

Add python:3.10
https://hub.docker.com/layers/python/library/python/3.10/images/sha256-0e6f1cd34a87218544d9e79858dfe68dd0f2d6919c2e2b287983276961aa4942?context=explore

meramsey committed:
Update ci.yml

add python:3.10
https://hub.docker.com/layers/python/library/python/3.10/images/sha256-0e6f1cd34a87218544d9e79858dfe68dd0f2d6919c2e2b287983276961aa4942?context=explore

meramsey committed:
Update pyproject.toml.jinja

Add missing dependency `safety` to resolve error running

```
@ poetry run duty check
✓ Checking code quality
✓ Type-checking
✓ Building documentation
✗ Checking dependencies (127)
  > poetry export -f requirements.txt --without-hashes | safety check --stdin --full-report
  /usr/bin/sh: 1: safety: not found
```

meramsey committed:
Update macros.py

Fix the set not being a list and then ignoring the other two lines as on a fresh install it fails currently

```
poetry run duty check
✓ Checking code quality
✗ Type-checking (1)
  > mypy --config-file config/mypy.ini src tests duties.py docs/macros.py
  docs/macros.py:34: error: Argument 1 to "search_packages_info" has incompatible type "Set[Any]"; expected "List[str]"
  docs/macros.py:35: error: Incompatible types in assignment (expression has type "Dict[Any, Union[str, List[str], None]]", variable has type "_PackageInfo")
  docs/macros.py:36: error: Invalid tuple index type (actual type "str", expected type "Union[int, slice]")
  Found 3 errors in 1 file (checked 8 source files)
```

meramsey commented on project/docs/macros.py:
```diff
@@ -31,9 +31,9 @@ def get_credits_data() -> dict:
-        pkg = {_: pkg[_] for _ in ("name", "home-page")}
-        packages[pkg["name"].lower()] = pkg
+    for pkg in search_packages_info(list(dependencies)):
+        pkg = {_: pkg[_] for _ in ("name", "home-page")}  # type: ignore
+        packages[pkg["name"].lower()] = pkg  # type: ignore
```
Not ideal but i think unless this is an easy fix should not be causing new users using this to start off with failing types when they are most likely not going to be experts in this area.

meramsey commented on project/docs/macros.py:
```diff
@@ -31,9 +31,9 @@ def get_credits_data() -> dict:
-    for pkg in search_packages_info(dependencies):
-        pkg = {_: pkg[_] for _ in ("name", "home-page")}
-        packages[pkg["name"].lower()] = pkg
+    for pkg in search_packages_info(list(dependencies)):
```
this one looked like easy cast to list so that is what I did here. Assuming should be fine unless it was supposed to be that way before.

meramsey committed:
Update multirun.sh

add python 3.10

meramsey committed:
Update setup.sh

add python 3.10

pawamoy commented on project/.github/workflows/ci.yml:
```diff
@@ -71,7 +71,7 @@ jobs:
       matrix:
         os: [ubuntu-latest, macos-latest, windows-latest]
-        python-version: [3.6, 3.7, 3.8, 3.9]
+        python-version: [3.6, 3.7, 3.8, 3.9, 3.10]
```
You'll need to wrap them in quotes, because in YAML 3.10 would be loaded as a float `3.1`.

pawamoy commented on project/.gitlab-ci.yml:
```diff
@@ -78,6 +81,7 @@ pages:
   - python3.7
   - python3.8
   - python3.9
+  - python:3.10
```
I think there's an extra colon here?

pawamoy commented on project/pyproject.toml.jinja:
```diff
@@ -28,6 +28,7 @@ pytest-cov = "^2.11.1"
 pytest-randomly = "^3.5.0"
 pytest-sugar = "^0.9.4"
 pytest-xdist = "^2.2.0"
+safety = "^1.10.3"
```
There's an issue with this: https://github.com/advisories/GHSA-7q25-qrjw-6fg2.
Unless you implement [the solution offered here](https://github.com/pyupio/safety/issues/348), safety should stay an external executable, for example installed with pipx. Feel free to document this instead.

pawamoy reviewed (changes requested):
Thanks!

Just a few things to fix.

meramsey committed:
Update .gitlab-ci.yml

meramsey committed:
Update pyproject.toml.jinja

meramsey commented on project/pyproject.toml.jinja:
```diff
@@ -28,6 +28,7 @@ pytest-cov = "^2.11.1"
 pytest-randomly = "^3.5.0"
 pytest-sugar = "^0.9.4"
 pytest-xdist = "^2.2.0"
+# safety = "^1.10.3" # install this locally outside of the venv `pip install safety --user`
```
What are your thoughts about leaving this inline comment so when people run into the error there aware of an easy safe solution. If its in documentation i missed then i would say we should remove it entirely but seems like something that might be helpful if someone is like why is this failing on first load if there is not a warning elsewhere when running `poetry run duty check` and it hits depends checks and notices its missing

pawamoy replied:
I think it would be better to print a warning in the `check-dependencies` duty when safety is not available.

```diff
     nofail = False
     safety = which("safety")
     if not safety:
         pipx = which("pipx")
         if pipx:
             safety = f"{pipx} run safety"
         else:
-            safety = "safety"
-            nofail = True
+            ctx.run("echo Please install safety or pipx to run this task && false", title="Checking dependencies", nofail=True)
+            return
```

meramsey replied:
@pawamoy 
As that is your code I took out the comment I had. Can you can add in your idea when doing next release?
I like that approach much better too

meramsey committed:
Update ci.yml

Ensure evaluated as string '3.10.0'  vs float 3.1

python:3.10.0

meramsey committed:
Update .gitlab-ci.yml

meramsey committed:
Update .pre-commit-config.yaml

bump mypy precommit hook to use one thats not buggy for python3.9

meramsey commented on project/.pre-commit-config.yaml:
```diff
@@ -14,7 +14,7 @@ repos:
         language_version: python3 # Should be a command that runs python3.6+
 -   repo: https://github.com/pre-commit/mirrors-mypy
-    rev: v0.720
+    rev: v0.910-1
```
This was needed cause i got errors from mypy precommit

```
@ pre-commit run --all-files
Trim Trailing Whitespace.................................................Passed
Fix End of Files.........................................................Passed
Check Yaml...............................................................Passed
Check for added large files..............................................Passed
black....................................................................Passed
mypy.....................................................................Failed
- hook id: mypy
- exit code: 2

duties.py:24: error: syntax error in type comment
duties.py:42: error: syntax error in type comment
duties.py:59: error: syntax error in type comment
duties.py:73: error: syntax error in type comment

Check pytest unit tests pass.............................................Passed
 v3.9.6 (.venv) ~/PycharmProjects/gitjirabot on  main [ 13📝 2● 1🤷 ⇡1] [✖ ERROR]  took 2s  at 14:19:55
 
@ 
```

After bumping to latest it now works :)
```
@ pre-commit run --all-files
[INFO] Initializing environment for https://github.com/pre-commit/mirrors-mypy.
[INFO] Installing environment for https://github.com/pre-commit/mirrors-mypy.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
Trim Trailing Whitespace.................................................Passed
Fix End of Files.........................................................Passed
Check Yaml...............................................................Passed
Check for added large files..............................................Passed
black....................................................................Passed
mypy.....................................................................Passed
Check pytest unit tests pass.............................................Passed
 v3.9.6 (.venv) ~/PycharmProjects/gitjirabot on  main [ 14📝 2● 1🤷 ⇡1]  took 7s  at 14:21:06
 
@

```

meramsey committed:
Update .pre-commit-config.yaml

Bump black version to one that supports python 3.10.0

meramsey commented:
I believe this should be all ready to go now @pawamoy 

Other than my docs pages not deploying for some reason on gitlab(probably user error on my part as all other stages complete) everything else seems fixed for stock issues I was running into when starting project.

Can see pipelines of the tests on my gitlab for recent project I started with same fixes from merge request
https://gitlab.com/mikeramsey/gitjirabot/-/pipelines

Thanks again for keeping this amazing project alive.

meramsey committed:
Update .pre-commit-config.yaml

Updated pre-commit version to latest

meramsey committed:
Update .gitlab-ci.yml

Updated to use main which is default branch on gitlab when creating repos

meramsey committed:
Update README.md.jinja

Fix default branch to main which it is on both github and gitlab now for new projects so that the badge images work. Also added the black style badge as it is in use by default.

meramsey committed:
Update pyproject.toml.jinja

remove uneeded comment about safety

pawamoy commented:
Alright lets do that. However note that I'm not at all using this template anymore, I've completely switched to PDM :slightly_smiling_face:  So thank **you** for maintaining it!

meramsey commented:
I appreciate you leaving it up at least its kinda my goto for new projects. I seen the PDM thing but haven't given it a try yet maybe i should to find out.
